### PR TITLE
Fix performance of Sabre rust<->Python boundary

### LINKED ISSUE #10650: Runtime Performance Regression in `SabreLayout`
### Environment

- **Qiskit Terra version**: 0.25.0
- **Python version**: 3.11
- **Operating system**: Linux

### What is happening?

When running `SabreLayout` with 0.25.0 has a large performance regression between 0.24.x and 0.25.0 that is a function of the number of gates in the circuit. The rust computation of the swap map and initial layout is as fast as before, but when we iterate over that to rewrite the `DAGCircuit` using that output from rust the runtime performance is significantly slower. For more modestly sized circuits this isn't super noticeable , but as the number of instructions in a circuit increases this overhead can be significantly higher.

### How can we reproduce the issue?

You can parse this qasm file from qasmbench: https://github.com/pnnl/QASMBench/blob/master/medium/bwt_n21/bwt_n21.qasm

```python
from qiskit import transpile, QuantumCircuit
from qiskit.providers.fake_provider import FakeCairoV2
import time
import logging
logging.basicConfig(level="INFO")

backend = FakeCairoV2()
for filename in ["bwt_n21.qasm"]:
    start = time.time()
    qc = QuantumCircuit.from_qasm_file(filename)
    parse_stop = time.time()
    print(f"parse {filename}: {parse_stop - start}")
    tqc = transpile(qc, backend, optimization_level=1)
    stop = time.time()
    print(f"{filename}: {stop - start}")
```

### What should happen?

The performance should be roughly equivalent to qiskit-terra < 0.25.0. Looking at 0.24.2 the above example `SabreLayout` took `22515.55800 ms` and 0.23.0 took `18759.72962 ms`. But I've been too impatient to wait for 0.25.0 to finish after several minutes.

### Any suggestions?

I expect the underlying issue stems from #10366 and specifically the rust<->python boundary doing unnecessary copies or conversions when we access data as we iterate over the circuit.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

In #10366 the SabreLayout and SabreSwap passes were refactored to support nested control flow blocks. As part of that refactor a new struct SabreResult was created to store the nested results for each block. This new class however resulted in PyO3 cloning the full swap map on every access. Since we have at least 1 access per gate in the circuit (and another one for each swap) this extra clone() adds a lot of extra overhead for deep circuits. In extreme cases this regression could be quite extreme. To address this the result format is changed to be a tuple (as it was originally), while this is less ergonomic the advantage it provides is that for nested objects it moves the rust object to the pyo3 interface so we avoid a copy as Python owns the object on the return. However, for control flow blocks we're still using the SabreResult class as it simplifies the internal implementation (which is why #10366 introduced the struct). The potential impact of this is mitigated by switching to only a single clone per control flow block, by only accessing the SabreResult object's attribute on each recursive call. However, if it is an issue in the future we can work on flattening the nested structure before returning it to python to avoid any clones.

### Details and comments

Fixes #10650

### CODE REVIEW

**Comment by mtreinish** [comment]
I reran the reproduce script from #10650 and with this PR applied `SabreLayout` took 18403.93066 ms which is inline with the 0.24.x and 0.23.x releases.

**Review by Eric-Arellano — APPROVED** [review]

File: qiskit/transpiler/passes/routing/sabre_swap.py [hunk 1]
```diff
@@ -341,11 +342,19 @@ def empty_dag(node, block):
         return out
 
     def apply_inner(out_dag, current_layout, qubit_indices_inner, result, id_to_node):
-        for node_id in result.node_order:
+        if isinstance(result, SabreResult):
```
Comment (kevinhartman) [inline qiskit/transpiler/passes/routing/sabre_swap.py:345]: You should be able to skip this check if you turn `block_result.result` into a tuple before this call.

https://github.com/Qiskit/qiskit-terra/pull/10652/files#diff-7d05140621ea52112a754c580b8d0056ed6cd8dcf44c7970ced6e2566959cc56R368-R374

**Review by kevinhartman — COMMENTED** [review]
Looks good, overall!

Thanks for the fix. Just one small comment that may be worth addressing.

File: qiskit/transpiler/passes/routing/sabre_swap.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/routing/sabre_swap.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10652/commits/bd734e216b9d23f7cb42bf50fc6672f0b70137a3

**Review by kevinhartman — APPROVED** [review]
LGTM!

**Pull request merged**